src/inmanta/env.py

```diff
@@ -314,11 +316,189 @@ class PipUpgradeStrategy(enum.Enum):
     ONLY_IF_NEEDED = "only-if-needed"
 
 
+class PipConfig(BaseModel):
+    """
+    :param use_config_file: Indicates whether the pip configuration files have to be taken into account when installing
+        Python packages.
+    :param index_url: one pip index url for this project.
+    :param extra_index_url:  additional pip index urls for this project. This is generally only
+        recommended if all configured indexes are under full control of the end user to protect against dependency
+        confusion attacks. See the `pip install documentation <https://pip.pypa.io/en/stable/cli/pip_install/>`_ and
+        `PEP 708 (draft) <https://peps.python.org/pep-0708/>`_ for more information.
+    :param pre:  allow pre-releases when installing Python packages, i.e. pip --pre.
+        Defaults to False. When false and pip.use-system-config=true we follow the system config.
+    :param use_system_config: defaults to false.
+        When true, sets the pip's index url, extra index urls and pre according to the respective settings outlined above
+            but otherwise respect any pip environment variables and/or config in the pip config file,
+            including any extra-index-urls.
+
+        If no indexes are configured in pip.index-url/pip.extra-index-url, this option falls back to pip's default behavior,
+        meaning it uses the pip index url from the environment, the config file, or PyPi, in that order.
+
+        For development, it is recommended to set this option to false, both for portability
+        (and related compatibility with tools like pytest-inmanta-lsm) and for security
+        (dependency confusion attacks could affect users that aren't aware that inmanta installs Python packages).
+    """
+
+    index_url: Optional[str] = None
+    # Singular to be consistent with pip itself
+    extra_index_url: Sequence[str] = []
+    pre: bool = False
+    use_system_config: bool = False
+
+    def has_source(self) -> bool:
+        """Can this config get packages from anywhere?"""
+        return self.index_url or self.use_system_config
+
+    def assert_has_source(self, reason: str) -> None:
+        if not self.has_source():
+            # TODO: Is this the correct exception type here?
+            # It is consistent with was there before, but is it Correct
+            # TODO: import cycle
+            from inmanta.module import InvalidModuleException
+
+            raise InvalidModuleException(
+                f"Attempting to install {reason} but pip is not configured. Add the relevant pip "
+                f"indexes to the project config file. e.g. to add PyPi as a module source, add the following "
+                "to `project.yml`:"
+                "\npip:"
+                "\n  index_url:"
+                "\n    - https://pypi.org/simple"
+                "\nAnother option is to set `pip.use_system_config = true` to use the system's pip config."
+            )
+
+
 class PipCommandBuilder:
```

Comment:
No longer the perfect name, but I'm afraid to rename it.
What should I do here? 
- Rename it
- Make a new superclass e.g. `PipHandler`
- Move this code elsewhere (into the config object or the venv object, but I don't feel either is a good choice)

Reply:
Any specific reason why you are afraid to rename it ?

Reply:
It has not stable API annotation, so perhaps I should just rename it

Reply:
Why did we add the `run_*` method to this class. This class had one very clearly responsibility. With this change we move away from that. I am not convinced that's a good idea.

Reply:
Don't these `run_` method belong in the `VirtualEnv` class? Because they act on a Python environment.

Reply:
I see why you did it and I think it makes sense. The only thing that still annoys me at the moment is that the `PythonEnvironment` class has a `_run_pip_install_command` method that does many thing that the new `run_*` methods in the `PipCommandBuilder` class do. Both take care of the logging, etc. So this more or less violates the objective of 1. Are you planning to move this code from the `PythonEnvironment` class to the `PipCommandBuilder` class as well?

Reply:
I did